lazy.go

```diff
@@ -69,13 +73,17 @@ func (l *lazyConn) Read(b []byte) (int, error) {
 }
 
 func (l *lazyConn) readHandshake() error {
-	l.rhlock.Lock()
-	defer l.rhlock.Unlock()
-
+	l.rhlock.RLock()
+	rhsync := l.rhsync
+	l.rhlock.RUnlock()
 	// if we've already done this, exit
-	if l.rhsync {
+	if rhsync {
 		return l.rerr
 	}
+
+	l.rhlock.Lock()
+	defer l.rhlock.Unlock()
+
 	l.rhsync = true
```

Comment:
This is racy. You can't check, unlock, then relock and expect the state to stay the same.